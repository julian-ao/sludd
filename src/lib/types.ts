export type LocationQueryData = {
    metadata?: {
        totaltAntallTreff: number;
    };
    navn: LocationData[];
};

export type LocationData = {
    representasjonspunkt: {
        nord: number;
        øst: number;
    };
    stedsnavn: {
        skrivemåte: string;
    }[];
    stedsnummer: string;
    navneobjekttype?: string;
    kommuner?: {
        kommunenummer?: string;
        kommunenavn?: string;
    }[];
};

export type WeatherQueryData = {
    properties: {
        meta: {
            updated_at: string;
        };
        timeseries: {
            time: string;
            data: {
                instant: {
                    details: {
                        air_temperature: number;
                        wind_speed: number;
                        relative_humidity: number;
                        wind_from_direction: number;
                    };
                };
                next_1_hours: {
                    summary: {
                        symbol_code: string;
                    };
                    details: {
                        precipitation_amount: number;
                    };
                };
            };
        }[];
    };
};
