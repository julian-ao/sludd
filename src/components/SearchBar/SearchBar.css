.searchContainer {
  width: 40%;
  height: 60px;
  display: flex;
  align-items: center;
  flex-direction: column;
  position: relative;
  margin-bottom: 80px;
}

.searchBar {
  width: 100%;
  display: flex;
  align-items: center;
  background-color: #fff;
  border-radius: 10px;
}

input[type="text"] {
  flex: 1;
  height: 60px;
  padding-left: 10px;
  border: none;
  outline: none;
  font-size: 20px;
  color: #777;
}

input[type="text"]::placeholder {
  color: #C7C7C7;
}

.searchDropdown {
  width: 100%;
  max-height: 350px;
  display: flex;
  flex-direction: column;
  background-color: #fff;
  border-top: 1px solid #C7C7C7;
  position: absolute;
  top: 60px;
  z-index: 1;
  border-bottom-left-radius: 10px;
  border-bottom-right-radius: 10px;
}

.searchDropdownItem {
  width: 100%;
  height: 50px;
  display: flex;
  align-items: center;
  color: #888;
  cursor: pointer;
  transition: background-color 0.3s;
}

.dropdownTextDiv {
  display: flex;
  flex-direction: column;
  margin-left: 12px;
}

.dropdownTextDiv h3 {
  margin: 0;
  margin-bottom: 1px;
  font-weight: 300;
}

.dropdownTextDiv p {
  margin: 0;
  font-size: 12px;
}

.selectedDropdownItem {
  background-color: #ddd;
}

.searchDropdownItem:hover {
  background-color: #ddd;
}

.searchButton {
  border: none;
  margin-right: 10px;
  background-color: transparent;
  cursor: pointer;
  transition: background-color 0.3s;
}

.searchButton:hover {
  cursor: pointer;
}

