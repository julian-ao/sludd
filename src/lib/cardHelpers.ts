// Dictionary of weather descriptions for the different weather types
const weatherDescription: { [key: string]: string } = {
    clearsky: "Clear sky",
    fair: "Fair",
    partlycloudy: "Partly cloudy",
    cloudy: "Cloudy",
    lightrainshowers: "Light rain showers",
    rainshowers: "Rain showers",
    heavyrainshowers: "Heavy rain showers",
    lightrainshowersandthunder: "Light rain showers and thunder",
    rainshowersandthunder: "Rain showers and thunder",
    heavyrainshowersandthunder: "Heavy rain showers and thunder",
    lightsleetshowers: "Light sleet showers",
    sleetshowers: "Sleet showers",
    heavysleetshowers: "Heavy sleet showers",
    lightssleetshowersandthunder: "Light sleet showers and thunder",
    sleetshowersandthunder: "Sleet showers and thunder",
    heavysleetshowersandthunder: "Heavy sleet showers and thunder",
    lightsnowshowers: "Light snow showers",
    snowshowers: "Snow showers",
    heavysnowshowers: "Heavy snow showers",
    lightssnowshowersandthunder: "Light snow showers and thunder",
    snowshowersandthunder: "Snow showers and thunder",
    heavysnowshowersandthunder: "Heavy snow showers and thunder",
    lightrain: "Light rain",
    rain: "Rain",
    heavyrain: "Heavy rain",
    lightrainandthunder: "Light rain and thunder",
    rainandthunder: "Rain and thunder",
    heavyrainandthunder: "Heavy rain and thunder",
    lightsleet: "Light sleet",
    sleet: "Sleet",
    heavysleet: "Heavy sleet",
    lightsleetandthunder: "Light sleet and thunder",
    sleetandthunder: "Sleet and thunder",
    heavysleetandthunder: "Heavy sleet and thunder",
    lightsnow: "Light snow",
    snow: "Snow",
    heavysnow: "Heavy snow",
    lightsnowandthunder: "Light snow and thunder",
    snowandthunder: "Snow and thunder",
    heavysnowandthunder: "Heavy snow and thunder",
    fog: "Fog",
};

// Takes a symbol_code and returns the corresponding weather description
export function parseWeatherDescription(symbol_code: string) {
    const splitSymbolCode = symbol_code.split("_");
    const description = weatherDescription[splitSymbolCode[0]];

    return description;
}
// Dictionary of colors for the different weather types
const weatherColors = {
    clearsky: ["#ded364", "#ede26f", "#d9cf64"],
    fair: ["#8FEFBE", "#90F0AE", "#90F0CF"],
    cloudy: ["#8C9C9A", "#8D9C8E", "#9C998C"],
    rain: ["#8AD2EA", "#A4DCEE", "#93d1e6"],
    snow: ["#6E8FF5", "#6EA7F5", "#6EBEF5"],
    sleet: ["#9FA1ED", "#AB9FEC", "#B89FED"],
    fog: ["#5E6385", "#605E84", "#675E85"],
};

// Takes a symbol_code and returns the corresponding color
export function getColorFromWeatherDescription(symbol_code: string) {
    //use weatherColors to get color from symbol_code
    for (const [key, value] of Object.entries(weatherColors)) {
        if (symbol_code.includes(key)) {
            if (Array.isArray(value)) {
                return value[Math.floor(Math.random() * value.length)];
            }
            return value;
        }
    }
    return "#000000";
}
