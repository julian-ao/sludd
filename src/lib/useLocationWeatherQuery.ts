import { useEffect, useState } from "react";
import { useQuery } from "@tanstack/react-query";
import { LocationQueryData, WeatherQueryData } from "./types";

interface locationName {
    locationName: string;
    filter?: string;
}

interface locationId {
    locationId: string;
    filter?: string;
}

// Custom hook for fetching weather data based on locationName or locationId
export const useLocationWeatherQuery = (data: locationName | locationId) => {

    const [locationData, setLocationData] = useState<LocationQueryData | null>(
        null,
    );

    // Fetch location data (cordinates) based on locationName or locationId
    const locationQuery = useQuery<LocationQueryData, unknown>(
        ["location", data],
        async () => {
            if ("locationId" in data && data.locationId) {
                const url = `https://ws.geonorge.no/stedsnavn/v1/sted?stedsnummer=${
                    data.locationId
                }&utkoordsys=4258${
                    data.filter ? `&filtrer=${data.filter}` : ""
                }`;
                const res = await fetch(url);
                return res.json();
            } else if ("locationName" in data) {
                const base_url = `https://ws.geonorge.no/stedsnavn/v1/sted?sok=${data.locationName}&utkoordsys=4258&treffPerSide=1&side=1&filtrer=navn.representasjonspunkt,navn.stedsnavn.skrivemåte,navn.navneobjekttype,navn.stedsnummer`;
                const res = await fetch(base_url);
                return res.json();
            }

            throw new Error("Invalid data");
        },
    );

    // Set locationData when locationQuery is successful
    useEffect(() => {
        if (locationQuery.isSuccess) {
            setLocationData(locationQuery.data!);
        }
    }, [locationQuery]);

    // Fetch weather data based on cords from locationData
    const weatherQuery = useQuery<WeatherQueryData, unknown>(
        ["weather", locationData?.navn[0].representasjonspunkt, location],
        async () => {
            const res = await fetch(
                `https://api.met.no/weatherapi/locationforecast/2.0/compact?lat=${locationData?.navn[0].representasjonspunkt.nord}&lon=${locationData?.navn[0].representasjonspunkt.øst}`,
            );
            return res.json();
        },
        {
            enabled: !!locationData,
        },
    );

    return { locationData, weatherQuery };
};
