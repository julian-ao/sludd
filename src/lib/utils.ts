// Convert date to readable format
export const convertDateToReadable = (date: string) => {
    const month = date.slice(5, 7);
    const day = parseInt(date.slice(8, 10), 10).toString(); // Remove leading zero
    const hour = date.slice(11, 13);
    const minute = date.slice(14, 16);

    return `${day}. ${months[
        Number(month) - 1
    ].toLocaleLowerCase()} ${hour}:${minute}`;
};

const months = [
    "Januar",
    "Februar",
    "Mars",
    "April",
    "Mai",
    "Juni",
    "Juli",
    "August",
    "September",
    "Oktober",
    "November",
    "Desember",
];
