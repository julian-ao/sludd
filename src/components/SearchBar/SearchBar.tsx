import React, { useState, useEffect, useRef } from 'react';
import { useQuery } from "@tanstack/react-query";
import { FaSearch, FaMapMarkerAlt } from 'react-icons/fa';
import './SearchBar.css';

const SearchBar: React.FC = () => {
    const [searchTerm, setSearchTerm] = useState<string>('');
    const [showDropdown, setShowDropdown] = useState<boolean>(false);
    const [selectedOptionIndex, setSelectedOptionIndex] = useState<number>(-1);
    const [lastKeyPressed, setLastKeyPressed] = useState<string>('');
    const searchBarRef = useRef<HTMLDivElement | null>(null);


    // Fetch data from API with searchTerm and fuzzy search = true
    const { data, refetch } = useQuery({
        queryKey: ["cordinatesData"],
        enabled: false,
        queryFn: () =>
            fetch(`https://ws.geonorge.no/stedsnavn/v1/navn?sok=${searchTerm}&fuzzy=true&utkoordsys=4258&treffPerSide=7&side=1`)
        .then((res) => res.json()),
    });

    // Fetch data from API and showDropdown when searchTerm changes
    useEffect(() => {
        if (searchTerm.trim() !== '') {0
            if (!['ArrowUp', 'ArrowDown', 'Enter'].includes(lastKeyPressed)) {
                refetch().then(() => setShowDropdown(true));
            }
        } else {
            setShowDropdown(false);
        }
    }, [searchTerm]);


    // Detect clicks outside of the dropdown and close it
    useEffect(() => {
        const handleClickOutside = (event: MouseEvent) => {
            if (searchBarRef.current && !searchBarRef.current.contains(event.target as Node)) {
                closeDropDown();
            }
        };
        document.addEventListener('click', handleClickOutside);
        return () => {
            document.removeEventListener('click', handleClickOutside);
        };
    }, []);


    // Handle search that calls onSearch function prop and close dropdown
    function handleSearch() {
        if (searchTerm.trim() === '') {
           alert("Mostl likely not a valid place in Norway")
        } else {
            closeDropDown();
            console.log("Search for: " + searchTerm);
        }
    }

    // Handle search with parameter that calls onSearch function prop and close dropdown
    function handleSearchWithParam(updatedSearchTerm: string) {
        setSearchTerm(updatedSearchTerm)
        closeDropDown();
        console.log("Search for: " + searchTerm);
    }


    // Close dropdown and reset selectedOptionIndex
    function closeDropDown() {
        setShowDropdown(false);
        setSelectedOptionIndex(-1);
    }


    // Handle key up events (Enter, ArrowUp, ArrowDown)
    const handleKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
        setLastKeyPressed(event.key);
        if (event.key === 'Enter') {
            handleSearch();
        } else if (['ArrowUp', 'ArrowDown'].includes(event.key) && showDropdown) {
            const itemCount = data?.navn?.length || 0;
            if (itemCount > 0) {
                let newIndex = (selectedOptionIndex + (event.key === 'ArrowUp' ? -1 : 1) + itemCount) % itemCount;
                newIndex = newIndex < 0 ? itemCount - 1 : newIndex;
                setSelectedOptionIndex(newIndex);
                setSearchTerm(data?.navn[newIndex].skrivemåte || '');
            }
        }
    };


    return (
        <div className='searchContainer' ref={searchBarRef}>
            <div
                className="searchBar"
                style={{
                    borderBottomLeftRadius: showDropdown ? 0 : '10px',
                    borderBottomRightRadius: showDropdown ? 0 : '10px'
                }}
            >
                <FaMapMarkerAlt size={18} color="#999" style={{paddingLeft: "10px"}} />
                <input
                  type="text"
                  placeholder="Søk på et sted i Norge..."
                  value={searchTerm}
                  onChange={(event) => setSearchTerm(event.target.value)}
                  onKeyUp={handleKeyPress}
                />
                <button onClick={handleSearch} className='searchButton'>
                    <FaSearch size={25} color="#999" />
                </button>
            </div>
            {showDropdown && data && (
                <div className='searchDropdown'>
                    {data.navn.map((item: any, index: number) => (
                        <div
                          className={`searchDropdownItem ${index === selectedOptionIndex ? 'selectedDropdownItem' : ''}`}
                          key={index}
                          onClick={() => {handleSearchWithParam(item.skrivemåte)}}
                        >
                            <FaMapMarkerAlt size={18} color="#999" style={{paddingLeft: "10px"}} />
                            <div className='dropdownTextDiv'>
                                <h3>{item.skrivemåte}</h3>
                                <p>{item.navneobjekttype}, {item.kommuner[0].kommunenavn}</p>
                            </div>
                        </div>
                    ))}
                </div>
            )}
        </div>
    );
};

export default SearchBar;
