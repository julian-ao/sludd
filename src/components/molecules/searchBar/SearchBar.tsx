import { useState, useEffect, useRef, useCallback } from 'react';
import { useQuery } from '@tanstack/react-query';
import { FaSearch, FaMapMarkerAlt } from 'react-icons/fa';
import { Link, useNavigate } from 'react-router-dom';
import { LocationQueryData } from '../../../lib/types';
import './searchBar.css';
import Popup from '../../atoms/popup/Popup';

export const API_URL = 'https://ws.geonorge.no/stedsnavn/v1/sted';

type SearchBarProps = {
    onSearch?: () => void;
};

const SearchBar = ({ onSearch }: SearchBarProps) => {
    const [searchTerm, setSearchTerm] = useState<string>('');
    const [showDropdown, setShowDropdown] = useState<boolean>(false);
    const searchBarRef = useRef<HTMLDivElement | null>(null);
    const [selectedOptionIndex, setSelectedOptionIndex] = useState<number>(-1);
    const [lastKeyPressed, setLastKeyPressed] = useState<string>('');
    const [isPopupShown, setIsPopupShown] = useState<boolean>(false);
    const navigate = useNavigate();

    // Fetch data from API based on search term, with fuzzy search true and hit limit 7
    const { data, refetch } = useQuery({
        queryKey: ['cordinatesData'],
        enabled: false,
        queryFn: () =>
            fetch(
                `${API_URL}?sok=${searchTerm}&fuzzy=true&utkoordsys=4258&treffPerSide=7&side=1`,
            ).then((res) => {
                if (!res.ok) {
                    throw new Error('Network response was not ok');
                }
                return res.json();
            }),
    });

    // Refetch data when search term changes,
    // then set showSropdown true  and reset selected option index
    useEffect(() => {
        if (searchTerm.trim() !== '') {
            if (!['ArrowUp', 'ArrowDown'].includes(lastKeyPressed)) {
                refetch()
                    .then(() => setShowDropdown(true))
                    .then(() => setSelectedOptionIndex(-1));
            }
        }
    }, [searchTerm]);

    // Handle click outside of search bar and set showDropdown false
    const handleClickOutside = useCallback((event: MouseEvent) => {
        if (
            searchBarRef.current &&
            !searchBarRef.current.contains(event.target as Node)
        ) {
            setShowDropdown(false);
            setSelectedOptionIndex(-1);
        }
    }, []);

    // Add event listener for click outside of search bar
    useEffect(() => {
        document.addEventListener('click', handleClickOutside);
        return () => {
            document.removeEventListener('click', handleClickOutside);
        };
    }, [handleClickOutside]);

    // Handle search if search term is not empty and navigate to search page
    const handleSearch = useCallback(() => {
        if (searchTerm.trim() === '') {
            showPopup();
        } else {
            setShowDropdown(false);
            if (onSearch) onSearch();
            navigate(`/search?q=${searchTerm}`);
        }
    }, [navigate, searchTerm]);

    // Handle key up events (Enter, ArrowUp, ArrowDown)
    const handleKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
        setLastKeyPressed(event.key);
        if (event.key === 'Enter') {
            if (selectedOptionIndex >= 0) {
                const selectedItem = data?.navn[selectedOptionIndex];
                navigate(
                    `/location/${selectedItem?.stedsnavn?.[0]?.skrivem친te}/${selectedItem?.stedsnummer}`,
                );
            } else {
                handleSearch();
            }
        } else if (
            ['ArrowUp', 'ArrowDown'].includes(event.key) &&
            showDropdown
        ) {
            const itemCount = data?.navn?.length || 0;
            if (itemCount > 0) {
                let newIndex =
                    (selectedOptionIndex +
                        (event.key === 'ArrowUp' ? -1 : 1) +
                        itemCount) %
                    itemCount;
                newIndex = newIndex < 0 ? itemCount - 1 : newIndex;
                setSelectedOptionIndex(newIndex);
                setSearchTerm(
                    data?.navn[newIndex].stedsnavn?.[0]?.skrivem친te || '',
                );
            }
        }
    };

    // Show popup for 1.5 seconds
    const showPopup = () => {
        setIsPopupShown(true);

        setTimeout(() => {
            setIsPopupShown(false);
        }, 1500);
    };

    return (
        <div className="searchContainer" ref={searchBarRef}>
            <Popup text={'游댌 Skriv inn et s칮keord'} show={isPopupShown} />
            <div
                className="searchBar"
                style={{
                    borderBottomLeftRadius:
                        showDropdown && data?.navn?.length > 0 ? 0 : '10px',
                    borderBottomRightRadius:
                        showDropdown && data?.navn?.length > 0 ? 0 : '10px',
                }}
            >
                <FaMapMarkerAlt
                    size={18}
                    color="#999"
                    style={{ paddingLeft: '10px' }}
                />
                <input
                    type="text"
                    placeholder="S칮k p친 et sted i Norge..."
                    value={searchTerm}
                    onChange={(event) => setSearchTerm(event.target.value)}
                    onKeyUp={handleKeyPress}
                />
                <button className="searchButton" onClick={handleSearch}>
                    <FaSearch size={25} color="#999" />
                </button>
            </div>
            {showDropdown && data && (
                <div
                    className="searchDropdown"
                    style={{
                        borderTop:
                            data?.navn?.length > 0
                                ? '1px solid #C7C7C7'
                                : 'none',
                    }}
                >
                    {(data as LocationQueryData)?.navn?.map(
                        (item, index: number) => (
                            <Link
                                className={`searchDropdownItem ${
                                    index === selectedOptionIndex
                                        ? 'selectedDropdownItem'
                                        : ''
                                }`}
                                key={index}
                                to={`/location/${item?.stedsnavn?.[0]?.skrivem친te}/${item?.stedsnummer}`}
                            >
                                <FaMapMarkerAlt
                                    size={18}
                                    color="#999"
                                    style={{ paddingLeft: '10px' }}
                                />
                                <div className="dropdownTextDiv">
                                    <h3>
                                        {item.stedsnavn?.[0]?.skrivem친te || ''}
                                    </h3>
                                    <p>
                                        {item.navneobjekttype}
                                        {item.kommuner?.[0]?.kommunenavn &&
                                            ', ' +
                                                item.kommuner?.[0]?.kommunenavn}
                                    </p>
                                </div>
                            </Link>
                        ),
                    )}
                </div>
            )}
        </div>
    );
};

export default SearchBar;
