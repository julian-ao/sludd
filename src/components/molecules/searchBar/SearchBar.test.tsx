import { cleanup, render } from '@testing-library/react'
import SearchBar from './SearchBar'
import { BrowserRouter } from 'react-router-dom'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { vi } from 'vitest';
import { LocationQueryData } from '../../../lib/types';
import React from 'react';

const wrapper = ({ children }: React.PropsWithChildren<object>) => (
    <BrowserRouter>
        {children}
    </BrowserRouter>
)

describe('SearchBar', () => {
    let queryClient: QueryClient;

    beforeEach(() => {
        queryClient = new QueryClient();
    });

    afterEach(() => {
        vi.restoreAllMocks();
        cleanup();
    });

    it('should render correctly', async () => {
        const mockResponse = {
            navn: [
                {
                    representasjonspunkt: {
                        nord: 1,
                        øst: 1,
                    },
                    stedsnavn: [
                        {
                            skrivemåte: 'Oslo',
                        },
                    ],
                    stedsnummer: '1',
                    kommuner: [
                        {
                            kommunenummer: '1',
                            kommunenavn: 'Oslo',
                        },
                    ],
                },
            ],
        } as LocationQueryData;
        vi.spyOn(global, 'fetch').mockResolvedValueOnce({
            json: vi.fn().mockResolvedValueOnce(mockResponse),
        } as never);


        const { container } = render(
            <QueryClientProvider client={queryClient}>
                <SearchBar />
            </QueryClientProvider>,
            { wrapper },
        );
        expect(container).toMatchSnapshot();
    });
});