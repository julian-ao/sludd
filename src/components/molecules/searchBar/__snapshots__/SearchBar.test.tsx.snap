// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > should render correctly 1`] = `
<div>
  <div
    class="searchContainer"
  >
    <div
      class="searchBar"
      style="border-bottom-left-radius: 10px; border-bottom-right-radius: 10px;"
    >
      <svg
        color="#999"
        fill="currentColor"
        height="18"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(153, 153, 153); padding-left: 10px;"
        viewBox="0 0 384 512"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M172.268 501.67C26.97 291.031 0 269.413 0 192 0 85.961 85.961 0 192 0s192 85.961 192 192c0 77.413-26.97 99.031-172.268 309.67-9.535 13.774-29.93 13.773-39.464 0zM192 272c44.183 0 80-35.817 80-80s-35.817-80-80-80-80 35.817-80 80 35.817 80 80 80z"
        />
      </svg>
      <input
        placeholder="Søk på et sted i Norge..."
        type="text"
        value=""
      />
      <button
        class="searchButton"
      >
        <svg
          color="#999"
          fill="currentColor"
          height="25"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(153, 153, 153);"
          viewBox="0 0 512 512"
          width="25"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
